veps/sig-compute/vmtemplate-proposal.md

```diff
@@ -0,0 +1,752 @@
+`kubevirt/kubevirt` repository, allowing to better organize the new code and to
+keep it separate from the core KubeVirt implementation. This means the feature
+will have its own operator, API server and controller.
+
+This design proposes using a new `kubevirt/virt-template` repository. The
+components in this shall be named `virt-template-operator`,
+`virt-template-controller`, `virt-template-api` and so on.
+
+Although the feature should be developed separately, it should still be deployed
+by the `virt-operator` in the core repository of KubeVirt, so that some changes
+to the core repository are still required to finally integrate the feature into
+KubeVirt. However, changes to the main repository should be kept to a minimum.
+
+A small downside to this approach is that `kubevirt/kubevirt `will still carry
+dependencies on the separate repository, e.g. within virtctl as there is
+currently no plugin mechanism we could use to extend the binary from outside the
+core kubevirt codebase. This also means that this feature must correspond to the
+KubeVirt release cycle.
+
+The following changes to the main repository are required:
+
+* Add FeatureGate and/or configurable for the feature to the `KubeVirt` CRD
+* Extend `virt-operator` to allow deployment of the feature (pulling in
+  `virt-template-operator` so that it can deploy the feature.)
+* Extend `virtctl` with the commands described below (pulling in
+  dependencies and not using a plugin mechanism.)
+
+### template.kubevirt.io/v1alpha1
+
+#### VirtualMachineTemplate
+
+A new `template.kubevirt.io` APIGroup will be introduced with an initial
+`v1alpha1` version being provided. The APIGroup will provide the
+`VirtualMachineTemplate` CRD heavily influenced by the `template.openshift.io`
+APIGroup and `Template` CRD provided downstream within OKD and OpenShift.
+However, instead of being a generic template for all kinds of CRDs, this CRD
+will only focus on templating of KubeVirt `VirtualMachines`.
+
+All features supported by the `VirtualMachine` CRD will be supported in the
+template, i.e. you can also use instance types and preferences in a template.
+Placeholders, such as `${NAME}`, are replaced with specific values when the
+template is processed.
+
+Deployment of the whole template feature will sit behind the `Template`
+FeatureGate in core KubeVirt.
+
+This initial version focuses on the use cases described above of the simple VM
+template based on a golden image and the creation of templates from existing
+VMs. The import and export use case of templates is to be implemented at a later
+stage.
+
+Initially, this will handle cross-namespace templates to VM creation using the
+same CDI workaround as the common-templates project currently employs. Providing
+a namespaced `SourceRef` reference to a `DataSource` golden image within the
+DataVolumeTemplate section of the VirtualMachine. This is currently the only way
+a VirtualMachine can be created using storage from another namespace
```

Comment:
A VM may have a persistent volume that is not a dataVolumeTemplate or even a DataVolume. How will those be handled? I think the process of creating the VirtualMachineTemplate will have to include injecting dataVolumeTemplates for each of those volumes in order for the cross namespace auth to work properly